File: src/main/java/com/github/zyypj/tadeuBooter/api/downloaders/dependencies/Dependencies.java
package com.github.zyypj.tadeuBooter.api.downloaders.dependencies;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.stream.JsonReader;
import com.github.zyypj.tadeuBooter.loader.Constants;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.bukkit.Bukkit;
import org.bukkit.plugin.Plugin;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.util.*;
import java.util.Map.Entry;
import java.util.concurrent.*;

/**
 * Essa classe foi inspirada numa implementação já existente!
 *
 * @author syncwrld (github.com/syncwrld)
 */
@Getter
@RequiredArgsConstructor
public class Dependencies {
    private final Queue<CompletableFuture<Void>> downloadQueue = new ConcurrentLinkedQueue<>();
    private final Plugin plugin;
    private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();

    /**
     * Cria uma instância de Dependencies associada ao plugin.
     *
     * @param plugin Plugin associado
     * @return Instância de Dependencies
     */
    public static Dependencies of(Plugin plugin) {
        return new Dependencies(plugin);
    }

    /**
     * Lê um arquivo JSON e converte-o em uma lista de dependências.
     *
     * @param file Arquivo JSON contendo as dependências
     * @return Array de dependências carregadas
     */
    public static Dependency[] of(File file) {
        List<Dependency> dependencies = new ArrayList<>();

        try (Reader reader = new FileReader(file);
             JsonReader jsonReader = new JsonReader(reader)) {

            JsonObject jsonObject = Constants.GSON.fromJson(jsonReader, JsonObject.class);

            for (Entry<String, JsonElement> entry : jsonObject.entrySet()) {
                JsonArray dependencyArray = entry.getValue().getAsJsonArray();
                if (dependencyArray.size() >= 3) {
                    DependencyInfo dependencyInfo = new DependencyInfo(
                            dependencyArray.get(0).getAsString(),
                            dependencyArray.get(1).getAsString(),
                            dependencyArray.get(2).getAsString()
                    );
                    dependencies.add(new Dependency(dependencyInfo));
                } else {
                    Bukkit.getConsoleSender().sendMessage("§4Dependência inválida no JSON: " + entry.getKey());
                }
            }
        } catch (IOException e) {
            Bukkit.getConsoleSender().sendMessage("§4Erro ao ler o arquivo de dependências: " + file.getAbsolutePath());
            Bukkit.getConsoleSender().sendMessage(e.getMessage());
        }

        return dependencies.toArray(new Dependency[0]);
    }

    /**
     * Adiciona uma dependência à fila de downloads.
     *
     * @param dependency Dependência a ser baixada
     */
    public void queue(Dependency dependency) {
        this.downloadQueue.add(dependency.download());
    }

    /**
     * Adiciona múltiplas dependências à fila de downloads.
     *
     * @param dependencies Dependências a serem baixadas
     */
    public void queue(Dependency... dependencies) {
        Arrays.stream(dependencies).forEach(this::queue);
    }

    /**
     * Remove uma dependência da fila de downloads.
     *
     * @param dependency Dependência a ser removida
     */
    public void remove(Dependency dependency) {
        this.downloadQueue.removeIf(future -> future.toString().equals(dependency.download().toString()));
    }

    /**
     * Executa a fila de downloads de forma assíncrona.
     */
    public void runQueue() {
        if (downloadQueue.isEmpty()) {
            Bukkit.getConsoleSender().sendMessage("§2Nenhuma dependência para baixar.");
            return;
        }

        CompletableFuture<Void> allDownloads = CompletableFuture.allOf(
                downloadQueue.toArray(new CompletableFuture[0])
        );

        allDownloads.join();
    }
}

File: src/main/java/com/github/zyypj/tadeuBooter/api/downloaders/dependencies/Dependencies.kt
package com.github.zyypj.tadeuBooter.api.downloaders.dependencies;

import com.github.zyypj.tadeuBooter.loader.Constants
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.stream.JsonReader
import org.bukkit.Bukkit
import org.bukkit.plugin.Plugin
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService

/**
 * Gerencia fila de downloads de dependências para um Plugin.
 */
class Dependencies private constructor(private val plugin: Plugin) {

    private val downloadQueue: Queue<CompletableFuture<Void>> = ConcurrentLinkedQueue()
    private val executor: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()

    companion object {
        /** Cria instância associada a um Plugin. */
        @JvmStatic
        fun of(plugin: Plugin): Dependencies = Dependencies(plugin)

        /** Carrega dependências de um JSON em arquivo. */
        @JvmStatic
        @Throws(IOException::class)
        fun of(file: File): Array<Dependency> {
            val list = mutableListOf<Dependency>()
            FileReader(file).use { fr ->
                JsonReader(fr).use { jr ->
                    val obj = Constants.GSON.fromJson<JsonObject>(jr, JsonObject::class.java)
                    for ((_, elem) in obj.entrySet()) {
                        val arr: JsonArray = elem.asJsonArray
                        if (arr.size() >= 3) {
                            val info = DependencyInfo(
                                arr[0].asString,
                                arr[1].asString,
                                arr[2].asString
                            )
                            list += Dependency(info)
                        } else {
                            Bukkit.getConsoleSender().sendMessage("§4Dependência inválida no JSON: $file")
                        }
                    }
                }
            }
            return list.toTypedArray()
        }
    }

    /** Enfileira download de uma dependência. */
    fun queue(dependency: Dependency) {
        downloadQueue.add(dependency.download())
    }

    /** Enfileira múltiplas dependências. */
    fun queue(vararg dependencies: Dependency) {
        dependencies.forEach { queue(it) }
    }

    /** Remove da fila. */
    fun remove(dependency: Dependency) {
        val id = dependency.dependencyInfo.name
        downloadQueue.removeIf { it.toString().contains(id) }
    }

    /** Executa a fila e aguarda conclusão. */
    fun runQueue() {
        if (downloadQueue.isEmpty()) {
            Bukkit.getConsoleSender().sendMessage("§2Nenhuma dependência para baixar.")
            return
        }
        CompletableFuture.allOf(*(downloadQueue.toTypedArray())).join()
    }
}